### ISSUE TITLE
all local and parl elections assume England & Wales rules

### ISSUE THREAD

**Issue by chris48s** [mrwilson/sopn-publish-date#1]
Hello. Just having a nose about :)

https://github.com/mrwilson/sopn-publish-date/blob/22acf993919915eed1049d7e3983f807bf6da297/sopn_publish_date/__init__.py#L14-L15

Assumes that all local and parl elections use the England & Wales rules, but in reality  I think `local.highland.wester-ross-strathpeffer-and-lochalsh.by.2018-12-06` should have a different SOPN date from `local.surrey.the-byfleets.by.2018-12-06` (for example) even though they are both `local.%.2018-12-06`.

In a case where the country/territory isn't implicit in the election type, it would be useful if it were possible to pass that in as a param.

**Comment by mrwilson** [comment]
That's a great point. I'm trying to accumulate edge-cases like this as tests.

I'm thinking an API change like:

```python
publish_date = sopn_publish_date(
  'local.highland.wester-ross-strathpeffer-and-lochalsh.by.2018-12-06',
  country='scotland'
)
```
Where `country` is default `None`.

**Comment by chris48s** [comment]
> I'm trying to accumulate edge-cases

You want edge cases.. I got edge cases :laughing: 

> I'm thinking an API change like...

Yep, so for the following election types, country is implicit:  `gla`, `naw`, `nia`, `sp`
and something like `sopn_publish_date('sp.c.aberdeen-central.2016-05-05', country='england')` would be complete nonsense.

Conversely, `local`, `parl`, `pcc` and in principle `mayor` are all non-country-specific so country should be required.

This makes the interface a bit clumsy because sometimes `country` is a required param, but sometimes it makes no sense to provide it. There is also the case where you have a group id like `parl.2017-06-08` which includes ballots in multiple countries.

Given all that, would it make more sense for `sopn_publish_date()` to just take a date and a country, as opposed to an election id? Is the ID actually used for anything other than extracting the date?

**Comment by mrwilson** [comment]
Re: why election id, my first attempt was to actually pass in an election type and date (as datetime) but I thought a single-input API was cleaner if it contained the same information.

However, as determining country from election location is _slightly_ out of scope for this project, taking a date and a country would be fine.

The important thing is utility.

**Comment by mrwilson** [comment]
From doing a bit of digging, it looks like it depends solely on country.
 * It's 23 working days in scotland (using the scottish public holiday calendar)
 * It's 19 working days in England/Wales (using E/W public holidays)
 * It's 19 working days in Northern Ireland (using NI public holidays)

Can you think of any other _weird_ edge cases? `pcc` elections follow the rules for country too, at a glance.

**Comment by mrwilson** [comment]
Found one! `gla` is 23 days.

https://londonelects.org.uk/im-candidate/statutory-notices-and-elections-timetable

**Comment by mrwilson** [comment]
Right, this has been resolved with some api changes. Given

```python
from sopn_publish_date import StatementPublishDate
from datetime import date

sopn_publish_date = StatementPublishDate()
```

* [`sopn_publish_date.local`](https://sopn-publish-date.readthedocs.io/en/latest/api.html#sopn_publish_date.StatementPublishDate.local) requires a non-optional `sopn_publish_date.calendars.Country` parameter

* Passing input of the form `local.YYYY-MM-DD` to [`sopn_publish_date.from_id`](https://sopn-publish-date.readthedocs.io/en/latest/api.html#sopn_publish_date.StatementPublishDate.from_id) will raise an `AmbiguousElectionId` exception.

* [`sopn_publish_date.uk_parliament`](https://sopn-publish-date.readthedocs.io/en/latest/api.html#sopn_publish_date.StatementPublishDate.uk_parliament) does not require a country (at the moment) because as far as I can see, modulo a handful of edge cases, the date is governed by a single piece of legislation, although maybe this is a function of historicdata not being tested across cross-country bank holiday edge-cases.

* Passing input of the form `parl.YYYY-MM-DD` to [`sopn_publish_date.from_id`](https://sopn-publish-date.readthedocs.io/en/latest/api.html#sopn_publish_date.StatementPublishDate.from_id) will not raise an exception.

**Comment by mrwilson** [comment]
Nope, I couldn't resist the temptation.

[`sopn_publish_date.uk_parliament`](https://sopn-publish-date.readthedocs.io/en/latest/api.html#sopn_publish_date.StatementPublishDate.uk_parliament) now takes an **optional** Country, with a default of `Country.ENGLAND`, with my reasoning explained in https://github.com/mrwilson/sopn-publish-date/commit/397dd2f9469ae9ee1548e93878151513e4b05091

**tl;dr** - governed by a _single_ piece of legislation rather than 2-4 so a difference in SoPN publish date results only from the differing calendars of the countries, and the default value of `Country.ENGLAND` passes 100% of historic tests anyway.

Gonna close this down ahead of 1.0.0
